packages/jest-environment-jsdom/src/index.ts

```diff
@@ -0,0 +1,146 @@
+      this.moduleMocker = null;
+      return;
+    }
+
+    // Node's error-message stack size is limited at 10, but it's pretty useful
+    // to see more than that when a test fails.
+    this.global.Error.stackTraceLimit = 100;
+    // `global` is of type `Win`, but `installCommonGlobals` expects
+    // `NodeJS.Global`, so using `any` for now
+    installCommonGlobals(global as any, config.globals);
+
+    // Report uncaught errors.
+    this.errorEventListener = event => {
+      if (userErrorListenerCount === 0 && event.error) {
+        (process as NodeJS.EventEmitter).emit('uncaughtException', event.error);
+      }
+    };
+    global.addEventListener('error', this.errorEventListener);
+
+    // However, don't report them as uncaught if the user listens to 'error' event.
+    // In that case, we assume the might have custom error handling logic.
+    const originalAddListener = global.addEventListener;
+    const originalRemoveListener = global.removeEventListener;
+    let userErrorListenerCount = 0;
+    global.addEventListener = function(name: string) {
+      if (name === 'error') {
+        userErrorListenerCount++;
+      }
+      // TODO: remove `any` type assertion
+      return originalAddListener.apply(this, arguments as any);
+    };
+    global.removeEventListener = function(name: string) {
+      if (name === 'error') {
+        userErrorListenerCount--;
+      }
+      // TODO: remove `any` type assertion
+      return originalRemoveListener.apply(this, arguments as any);
+    };
+
+    // `global` is of type `Win`, but `ModuleMocker` expects `NodeJS.Global`, so
+    // using `any` for now
+    this.moduleMocker = new mock.ModuleMocker(global as any);
+
+    const timerConfig = {
+      idToRef: (id: number) => id,
+      refToId: (ref: number) => ref,
+    };
+
+    this.fakeTimers = new FakeTimers({
+      config,
+      // `global` is of type `Win`, but `FakeTimers` expects `NodeJS.Global`, so
+      // using `any` for now
+      global: global as any,
+      moduleMocker: this.moduleMocker,
+      timerConfig,
+    });
+  }
+
+  setup() {
+    return Promise.resolve();
+  }
+
+  teardown() {
+    if (this.fakeTimers) {
+      this.fakeTimers.dispose();
+    }
+    if (this.global) {
+      if (this.errorEventListener && isWin(this.global)) {
+        this.global.removeEventListener('error', this.errorEventListener);
+      }
+      // Dispose "document" to prevent "load" event from triggering.
+      Object.defineProperty(this.global, 'document', {value: null});
+      if (isWin(this.global)) {
+        this.global.close();
+      }
+    }
+    this.errorEventListener = null;
+    this.global = null;
+    this.dom = null;
+    this.fakeTimers = null;
+    return Promise.resolve();
+  }
+
+  runScript(script: Script) {
+    if (this.dom) {
+      // Explicitly returning `unknown` since `runVMScript` currently returns
+      // `void`, which is wrong
+      return this.dom.runVMScript(script) as unknown;
```

Comment:
```suggestion
      return this.dom.runVMScript(script) as any;
```

We know what the type here will be, no need to assert it.

(it'll have the `EVAL_RESULT_VARIABLE` object thing due to https://github.com/facebook/jest/blob/438a178c8addf2806bebf44726d080921ad9984f/packages/jest-transform/src/ScriptTransformer.ts#L559-L565)

Reply:
We do need to assert it since `runVMScript()` returns `void`, which is not true, and `runScript()` is also expected to receive `{[ScriptTransformer.EVAL_RESULT_VARIABLE]: ModuleWrapper} | null` (via the JestEnvironment interface). I have a new solution for this, committing in a sec.

Reply:
This contrasts with https://github.com/facebook/jest/pull/8003#discussion_r260900500 - we can't both set the return type to the `EVAL_RESULT_VARIABLE` thing, AND the return value of the provided script... Sorry, I'm a bit confused here.

If we want the return type of `runScript` to be the the same as the return type of the provided script, we could use a generic to let the user of `jest-environment-jsdom` set the return type? Just an idea. But I feel there is something I am missing here... :)

Reply:
The return value of the provided script is, 100%, the `EVAL_RESULT_VARIABLE` thing. We construct the script ourselves. The function assigned to `EVAL_RESULT_VARIABLE`, we do not know what will return when invoked (that's the `unknown` part), but we don't care about that result

>If we want the return type of `runScript` to be the the same as the return type of the provided script, we could use a generic to let the user of `jest-environment-jsdom` set the return type? Just an idea. But I feel there is something I am missing here... :)

I don't think that's needed - the script we get is constructed from a string source we've read (and potentially transformed) from disk. So we cannot statically know what it will return